### PR TITLE
Issue-4: Add Either

### LINKED ISSUE #4: Add Either
`Either` is a right-biased disjunction that carries one of two types of value (such as `Exception` on the left, and `int` on the right, for example).  It fulfills the same categories as `Maybe`, but I just can't even begin to stand thinking about that.

### PR DESCRIPTION
**Pull request by doomchild**
Closes #4
enhancement
Semver:Minor

### CODE REVIEW

File: Highbar.Types/Either.cs [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+  {
+    /// <summary>
+    /// Captures the possible <c>Exception</c> in a <c>Left</c>, otherwise the <c>value</c> in a <c>Right</c>.
+    /// </summary>
+    public static Either<Exception, R> Attempt(Func<R> supplier)
+    {
+      try
+      {
+        return Right<Exception, R>(supplier());
+      } catch (Exception exception)
```
Comment (jlmorgan) [inline Highbar.Types/Either.cs]: Seems weird to cuddle the `catch` but not the braces… o_O

Comment (doomchild) [reply Highbar.Types/Either.cs]: Whoops, missed that one.

File: Highbar.Types/Either.cs [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+      return result;
+    }
+
+    /// <summary>
+    /// Creates an <c>Either</c> from a <c>Maybe</c>.
+    /// </summary>
+    public static Either<Exception, SR> From<SR>(Maybe<SR> maybe)
+    {
+      return (maybe == null)
+        ? Left<Exception, SR>(new NullReferenceException("maybe is null"))
```
Comment (jlmorgan) [inline Highbar.Types/Either.cs]: `maybe is null` or `maybe must not be null`?

Comment (doomchild) [reply Highbar.Types/Either.cs]: It's `must not` in other places.  I'll change it.

File: Highbar.Types/Either.cs [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+    /// Applies the <c>predicate</c> to the <c>Right</c> value.
+    /// </summary>
+    public abstract Either<L, R> Filter(Predicate<R> predicate, Func<L> supplier);
+
+    /// <summary>
+    /// Bifunctor <c>first</c>.  Maps the <c>Left</c> value if the <c>instance</c> is a <c>Left</c>.
+    /// </summary>
+    public Either<ML, R> First<ML>(Func<L, ML> mapper)
+    {
+      return Bimap(mapper, i => i);
```
Comment (jlmorgan) [inline Highbar.Types/Either.cs:220]: Is there not a static `identity` function? If not, consider implementing on like in `Functions` class from the `functional-extras` library.

Comment (doomchild) [reply Highbar.Types/Either.cs:220]: If there is, I don't know about it.  It's on the list of things I need to add (along with `Predicates` and `Suppliers`).

File: Highbar.Types/Either.cs [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+    /// Executes the <c>consumer</c> if the <c>instance</c> is a <c>Right</c>.
+    /// </summary>
+    public abstract Either<L, R> IfRight(Action<R> consumer);
+
+    /// <summary>
+    /// Determines whether or not the <c>instance</c> is a <c>Left</c>.
+    /// </summary>
+    public abstract bool IsLeft
+    {
+      get;
```
Comment (jlmorgan) [inline Highbar.Types/Either.cs:286]: Could do:

```csharp
return !isRight();
```

File: Highbar.Types/Either.cs [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+
+    /// <summary>
+    /// Recover from a <c>Left</c> into a possible <c>Right</c>.
+    /// </summary>
+    public abstract Either<L, R> Recover(R value);
+
+    /// <summary>
+    /// Recover from a <c>Left</c> into a <c>Right</c> with a value produced by a <c>Func</c>.
+    /// </summary>
+    public abstract Either<L, R> Recover<A>(Func<A, R> other);
```
Comment (jlmorgan) [inline Highbar.Types/Either.cs:318]: Is this suppose to be `Func<A, R>` and not `Func<L, R>`?

Comment (doomchild) [reply Highbar.Types/Either.cs:318]: It matches what's in the Java library.  It's a function where we don't really care what the input argument type is.

File: Highbar.Types/Nothing.cs [hunk 6]
```diff
@@ -119,9 +124,9 @@ public override int GetHashCode()
       return 1;
     }
 
-    public override bool Equals(object o)
+    public override bool Equals(object obj)
```
Comment (jlmorgan) [inline Highbar.Types/Nothing.cs:127]: Instead of an abbreviation, you could do `other`.

File: Highbar.Types/Left.cs [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+    public override Either<L, R> Alt(Func<L, Either<L, R>> other)
+    {
+      Objects.RequireNonNull(other, "other must not be null");
+
+      return other(default(L));
+    }
+
+    public override Either<L, MR> Ap<MR>(Either<L, Func<R, MR>> other)
+    {
+      return new Left<L, MR>(Either<L, MR>.FromLeft(this));
```
Comment (jlmorgan) [inline Highbar.Types/Left.cs]: After doing some verification on this particular method, I have found the following:

```haskell
instance Applicative (Either e) where
    pure          = Right
    Left  e <*> _ = Left e
    Right f <*> r = fmap f r
```

To translate, the value on the left of the `<*>` is the `other` in this context. So, regardless of what the current instance is, if the `other` is a `Left`, the `other` is returned.

Comment (doomchild) [reply Highbar.Types/Left.cs]: Fixed.

**Pull request merged**